### ISSUE TITLE
Whole Genome Runtime Regression

### ISSUE THREAD

**Issue by glennhickey** [ComparativeGenomicsToolkit/cactus#157]
I wanted to try a big graph before [merging a change to the ktserver version](#148).  So I ran a human-chimp alignment with gorilla outgroup that I'd previously tested on kubernetes.  When I tested it then (on python 2.7), it I think took about 10 hours on 80 cores.

Running now, it's going on about 20 hours on 90 cores and still hasn't finished the lastz stage (so no ktservers have even been made yet).   
```
%Cpu(s): 92.7 us,  0.5 sy,  0.0 ni,  6.7 id,  0.1 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem : 39590361+total, 34334259+free, 41878356 used, 10682680 buff/cache
KiB Swap:        0 total,        0 free,        0 used. 34738518+avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                   
20783 root      20   0  264440 175508   2592 R 100.0  0.0   1:27.64 cPecanLastz --format=cigar --notrivial --step=3 --ambiguous=iupac,100,100 --ydrop=3000 /+ 
22206 root      20   0  149432 130264   3032 R 100.0  0.0   0:23.45 cPecanRealign --gapGamma 0.0 --matchGamma 0.9 --diagonalExpansion 4 --splitMatrixBiggerT+ 
22379 root      20   0  263924 175324   2568 R 100.0  0.0   0:14.77 cPecanLastz --format=cigar --notrivial --step=3 --ambiguous=iupac,100,100 --ydrop=3000 /+ 
22556 root      20   0  246836 158200   2500 R 100.0  0.0   0:10.34 cPecanLastz --format=cigar --notrivial --step=3 --ambiguous=iupac,100,100 --ydrop=3000 /+ 
22579 root      20   0  263924 175288   2524 R 100.0  0.0   0:09.26 cPecanLastz --format=cigar --notrivial --step=3 --ambiguous=iupac,100,100 --ydrop=3000 /+ 
22647 root      20   0  263924 175092   2588 R 100.0  0.0   0:06.67 cPecanLastz --format=cigar --notrivial --step=3 --ambiguous=iupac,100,100 --ydrop=3000 /+ 
18308 root      20   0  390852 373012   3032 R  99.4  0.1   3:11.75 cPecanRealign --gapGamma 0.0 --matchGamma 0.9 --diagonalExpansion 4 --splitMatrixBiggerT+ 
20060 root      20   0  283560 196288   2580 R  99.4  0.0   1:54.52 cPecanLastz --format=cigar --notrivial --step=3 --ambiguous=iupac,100,100 --ydrop=3000 /+ 
20570 root      20   0  264440 175532   2572 R  99.4  0.0   1:36.47 cPecanLastz --format=cigar --notrivial --step=3 --ambiguous=iupac,100,100 --ydrop=3000 /+ 
etc. etc.
```
I'm not sure what's slowing it down.  

Could it be the disabling of [ChildTreeJob](https://github.com/ComparativeGenomicsToolkit/cactus/commit/9e41f0328d66b903d639286e16d85acb38d1110d)?   CPU utilization looks good, though, which isn't what I'd expect if this were an issue.  Still, [there's a patch to try](#147).

Is it related to the config.xml issue (#151)?  Are one of those few differences somehow slowing down the lastz stage?  Seems very unlikely  but easy enough to check. 

Is there something changed in the build itself, either in #148 or the py3 merge that's making lastz slower?   Or a subtle bug introduced along the way that's caused more alignments to be done?  I never picked up a runtime difference in the evolver tests, but the lastz phase only takes a few seconds in that.

**Comment by glennhickey** [comment]
My time estimates above fairly off.  Going back over the logs, when it ran through, there were 6228 `cPecanLastz --ungapped` invocations that happened over 13 hours.  On the current run, it's 2828 so far over the past 11 hours.  But I think the point still stands: it's on pace to be much slower for this phase when it should be faster (10% more cores, and I'm using local binaries instead of singularity).  On the old run, these commands look like
```
cPecanLastz tmpS2wAFk.tmp[multiple][nameparse=darkspace] tmpcv1Fds.tmp[nameparse=darkspace] --step=3 --ambiguous=iupac,100,100 --ungapped --queryhsplimit=keep,nowarn:1500 --querydepth=keep,nowarn:13 --format=general:name1,zstart1,end1,name2,zstart2+,end2+ --markend
```
and on the new run
```
cPecanLastz tmpfp_2t_4w.tmp[multiple][nameparse=darkspace] tmpe0gmgt5j.tmp[nameparse=darkspace] --step=3 --ambiguous=iupac,100,100 --ungapped --queryhsplimit=keep,nowarn:1500 --querydepth=keep,nowarn:4 --format=general:name1,zstart1,end1,name2,zstart2+,end2+ --markend
```
There is indeed a difference. 
```
Old: --querydepth=keep,nowarn:13
New: --querydepth=keep,nowarn:4
```

This parameter is coming from [LastzRepeatMaskJob](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/src/cactus/preprocessor/lastzRepeatMasking/cactus_lastzRepeatMask.py#L56-L77)

**Comment by glennhickey** [comment]
The `queryDepth` comes from the [line](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/src/cactus/preprocessor/cactus_preprocessor.py#L196):
```
self.period = max(1, round(self.proportionSampled * self.minPeriod))
```
which is coming up with 1 now, but 10 back in python2.7.  `self.minPeriod` looks to come directly from the XML as 50.  `self.proportionSampled` is from [here](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/src/cactus/preprocessor/cactus_preprocessor.py#L158):
```
outChunkIDList.append(self.addChild(self.getChunkedJobForCurrentStage(inChunkIDs, float(inChunkNumber)//len(inChunkIDList), inChunkIDList[i])).rv())
```
And that line above was indeed changed in the Python3 conversion. In particular  `float(inChunkNumber)/len(inChunkIDList)` was changed to `float(inChunkNumber)/len(inChunkIDList)`.  This clamps that fraction down to 0, leading to the different period. 

I guess this is something from `2to3`?  Luckily there only seem to be a handful of such cases, so it should be quick to verify each one. 

There's also a bug in this related [line](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/src/cactus/preprocessor/cactus_preprocessor.py#L196) where the default is a string instead of int, but that would presumably cause a crash if it came up so I don't think it's at all linked. 
```
minPeriod = getOptionalAttrib(prepNode, "minPeriod", typeFn=int, default="0")
```

**Comment by diekhans** [comment]
python 2 would have round '/' down to an int, so the behavior is the same.

This could be related to the problem of the multiple XMLs and config not being passed through.  I am working on that now.

**Comment by glennhickey** [comment]
I may have jumped the gun on complaining that it's slower.  Maybe only slower than I remembered.  Will check back once it completes.  

The difference in blast parameters *is* from changing `/` to `//` as described above.  That's corrected in #160.  Which is a good thing, but I am no longer convinced it's radically affected the runtime.

**Comment by glennhickey** [comment]
The rounding error was fixed in #160, and it looks like the runtime regression was all in my head.  It still won't run whole-genome, but hopefully #165 will take care of that.
